### PR TITLE
Dual quaternions

### PR DESCRIPTION
**Pull request by nanjizal**
Just current Dual Quaternion changes

### CODE REVIEW

**Comment by nanjizal** [comment]
@see https://github.com/tbrosman/hxmath/pull/64

**Comment by nanjizal** [comment]
Still WIP no additional testing yet added.

**Comment by nanjizal** [comment]
Stuff requested I action:

> Cool, I'll try to add more documentation/references when I get time to work on this again. This is one of the more useful papers I found while refreshing my brain: https://www.cs.utah.edu/~ladislav/kavan06dual/kavan06dual.pdf It goes over how to set up the product for transforming points.
> 
> In the meantime there are a few things this PR needs:
> 
> Code style should match the rest of the library
> Test for Quaternion.fromYawPitchRoll
> Test for DualQuaternion.fromAxisAngle, fromTranslation (should match result of getTranslation())
> Test for DualQuaternion.identity
> One more thing I keep forgetting: if I include your name in contributors, haxelib associates the library with your username. It is easier for me to just credit you in the CHANGELOG.md (and class file). I found this out the hard way a year ago when someone updated my Haxe REST Client library and I tried to submit the change. Basically haxelib doesn't differentiate contributors and owners: https://lib.haxe.org/documentation/faq/

**Comment by nanjizal** [comment]
Getting a strange edge case which I have omitted for now.
https://github.com/tbrosman/hxmath/pull/64/commits/eee23239d7e148a4197b972ca26cfbb2f3268063#diff-d019ead88319c7f5c2e86f45914ddc77R163
Still some more tests to add.  For moment I have put all Quaternions stuff in new Test class but can move it back to Test3D once I am happier with it.

**Comment by nanjizal** [comment]
@tbrosman Are these tests suitable or does it need more coverage? I am not yet sure how to deal with the one odd failure, I expect styling needs improvement I can move it back into Test3D and tidy styling if you feel the tests are suitable, and take another look at the edge case.

**Comment by nanjizal** [comment]
Had quite a few problems with pushing to specific branches, but my current changes should all be there.

**Comment by nanjizal** [comment]
I was going to ask do you use a plugin or something to apply styles, do you use the 'recommended' haxe styling, I think that has a plugin or similar but not tried it yet.

File: Test.hx [hunk 1]
```diff
@@ -3,10 +3,9 @@ import test.TestAll;
 class Test {
     static function main() {
         var passed = TestAll.main();
-        
         #if (sys && !EXIT_ON_FINISH)
         Sys.stdin().readLine();
-        #else
+        #elseif !js
```
Comment (tbrosman) [inline Test.hx:8]: Should be `#elseif sys`. I missed that case when I wrote the test driver.

File: hxmath/math/DualQuaternion.hx [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+    /**
+     * Construct a DualQuaternion using axis angle 
+     * axis angle in degrees and axis vector3, translation in vector4
+     * 
+     * @param angleDegrees  The angle to rotate in degrees.
+     * @param axis          The axis to rotate around.
+     * @param t             The Vector4 ( translation )
+     * @return              The DualQuaternion
+     */
+    public static inline function fromAxisAngle( angleDegrees:Float
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:96]: Minor: should match existing function style (in this case just put it all on the same line)

Spacing below should also match existing.

File: hxmath/math/DualQuaternion.hx [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+                                                , axis: Vector3
+                                                , vec4: Vector4 ):DualQuaternion
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:98]: Probably should be "t" or "translation" or something (also update documentation to match).

File: hxmath/math/DualQuaternion.hx [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+     * @param t             The Vector4 ( translation )
+     * @return              The DualQuaternion
+     */
+    public static inline function fromAxisRadian( theta:Float
+                                                , axis: Vector3
+                                                , vec4: Vector4 ):DualQuaternion 
+    {
+        var cosHalfAngle = Math.cos(theta / 2.0);
+        var sinHalfAngle = Math.sin(theta / 2.0);
+        var real = new Quaternion(cosHalfAngle,sinHalfAngle * axis.x,sinHalfAngle * axis.y,
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:117]: This should live in Quaternion. Then you can call Quaternion.fromAxisRadian.

File: hxmath/math/DualQuaternion.hx [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+        return Quaternion.dot(a.real.clone(), b.real.clone());
+    }
+    
+    /**
+     * Create a complex conjugate copy of a dualquaternion
+     * 
+     * @param a
+     * @return       ~a
+     */
+    @:op(~A)
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:201]: Like I said in my previous comment, this might not make the most sense for choice of conjugate. Here's my reasoning:

* Rotation by a unit length quaternion q is calculated as q * v * (~q) (where ~q is the complex conjugate).
* With this convention, transformation by a dual quaternion q would be written q * v * (~q.dualConjugate()) which is kind of gross (also, dualConjugate doesn't exist yet, so you can't actually write this).

File: hxmath/math/DualQuaternion.hx [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+    /**
+     * applyScale
+     * 
+     * @param a
+     * @return       scaled 
+     */
+    public inline function applyScale(s:Float):DualQuaternion
+    {
+        var cloned = clone();
+        cloned = cloned*DualQuaternion.scalar( s );
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:289]: Why not just call scaleMultiply? Unless scalar() is inlined, this will create an intermediate DualQuaternion.

File: hxmath/math/DualQuaternion.hx [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+    }
+    
+    private static inline function get_identity():DualQuaternion
+    {
+        return new DualQuaternion(Quaternion.identity, Quaternion.zero);
+    }
+    
+    private inline function get_normal():DualQuaternion
+    {
+        var mag = dot(this, this);
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:369]: This isn't actually the magnitude, but the magnitude squared. Just call .length.

File: hxmath/math/DualQuaternion.hx [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+        if( mag < 0 ) return null;
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:370]: This condition is impossible to hit. Dot products are always >= 0 by definition (if it isn't, it's not a dot product). I think you want to check for when mag == 0. In that particular case you should just return the existing dual quaternion (it is the zero dual quaternion).

File: hxmath/math/DualQuaternion.hx [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+    
+    /**
+     * Normalize the quaternion in-place.
+     * 
+     * @return  The modified object.
+     */
+    public inline function normalize(): DualQuaternion
+    {
+        var mag = dot( this, this );
+        if( mag < 0 ) return null;
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:383]: Duplicates code with get_normal. Suggestion: call .clone().normalize() in get_normal. I actually don't do this correctly in my Quaternion code, so normal does fail on 0-length quaternions (oops).

File: hxmath/math/DualQuaternion.hx [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+    
+    /**
+     * Rotate the given vector, assuming the current DualQuaternion is normalized (if not, normalize first).
+     * 
+     * @param u     The vector to rotate.
+     * @return      The rotated vector.
+     */
+    public inline function rotate(u:Vector3):Vector3
+    {
+        normalize();
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:570]: Quaternion doesn't implicitly normalize on rotate, so neither should DualQuaternion. It is up to the caller to make the trade-off between precision and speed.

File: hxmath/math/DualQuaternion.hx [hunk 11]
```diff
@@ -0,0 +1,622 @@
...
+    }
+    
+    
+    /**
+     * Find the arccosine of the angle between two rotation quaternions in the DualQuaternion
+     * 
+     * @param b     The other quaternion.
+     * @return      The arccosine angle between this vector and the other in radians.
+     */
+    public inline function angleWith(b:Quaternion):Float
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:581]: This seems unnecessary. There's an angleWith on Quaternion, so the caller can just call dualQuaternion.real.angleWith(quaternion).

File: hxmath/math/DualQuaternion.hx [hunk 12]
```diff
@@ -0,0 +1,622 @@
...
+        return 2.0 * Math.acos(Quaternion.dot(this.real, b) / (this.real.length * b.length));
+    }
+    
+    
+    /**
+     * Dual Conjugate the Dual Quaternion in-place.
+     * 
+     * @return  The modified object.
+     */
+    public inline function applyDualConjugate():DualQuaternion
```
Comment (tbrosman) [inline hxmath/math/DualQuaternion.hx:592]: Reminder: needs to be implemented.

File: hxmath/math/Quaternion.hx [hunk 13]
```diff
@@ -32,6 +32,38 @@ class QuaternionDefault
     }
 }
 
+// yaw (Z), pitch (Y), roll (X)
+@:forward
+abstract Euler( Vector3 ) from Vector3 to Vector3 {
+    public inline function new( roll: Float, pitch: Float, yaw: Float ){
```
Comment (tbrosman) [inline hxmath/math/Quaternion.hx:38]: Minor: order fields like existing structures (properties on top, make setter/getter implementations private and put them on the bottom).

File: test/TestStress.hx [hunk 14]
```diff
@@ -3,7 +3,7 @@ import hxmath.math.Matrix3x2;
 import hxmath.math.Matrix3x3;
 import hxmath.math.Matrix4x4;
 import hxmath.math.Vector2;
-
+#if !js
```
Comment (tbrosman) [inline test/TestStress.hx:6]: Should be `#if sys`

File: test_all_mac.hxml [hunk 15]
```diff
@@ -0,0 +1,36 @@
+-cmd echo 'js test'
```
Comment (tbrosman) [inline test_all_mac.hxml:1]: Is it not possible to reference hxml here? I admit I've never run this on a Mac. What's the difference? I see a lot of cd'ing into directories that doesn't happen on the "regular" hxml targets. Is it something to do with the path convention for running binaries/scripts?

**Comment by tbrosman** [comment]
Looking much better. I left a few more comments, mostly cleanup stuff (and the conjugate operator needs some work).

I'm not using haxecheckstyle or anything (I really should). One day I'll learn how to configure it to look the way I want it to.
